File: maven-simplest/src/main/python/ETL/TextProcessor.py
'''
@date Oct 20, 2015
@author: Hechao Li
'''
import re

class TextProcessor:
    # key: banned word; Value: asterisks 
    banned_words = {}
    sentiment_dict = {}
    
    def ROT13(self,text):
        result = ''
        for c in text:
            if c.isupper():
                c = chr((ord(c) - ord('A') + 13) % 26 + ord('A'))
            elif c.islower():
                c = chr((ord(c) - ord('a') + 13) % 26 + ord('a'))
            result += c
        return result
    
    def SentimentScore(self, text):
        sentiment = 0
        words = re.findall(r'\w+', text)
        for word in words:
            lower_word = word.lower()
            if lower_word in self.sentiment_dict:
                sentiment += self.sentiment_dict[lower_word]
        return sentiment
    
    def TextCensoring(self, text):
        words = re.findall(r'\w+', text)
        for word in words:
            lower_word = word.lower()
            if lower_word in self.banned_words:
                # Do not check if len < 2 because we trust banned.txt file
                censored_word = word[0] + self.banned_words[lower_word] + word[-1]
                text = re.sub(r'\b' + word + r'\b', censored_word, text)
        return text
    
    # Load file banned.txt and afinn.txt
    def __init__(self):
        with open('banned.txt') as f:
            for line in f:
                banned_word = self.ROT13(line.rstrip())
                censored_word = '*' * (len(banned_word)-2)
                self.banned_words[banned_word] = censored_word
        with open('afinn.txt') as f:
            for line in f:
                word, score = line.rstrip('\n').split('\t')
                self.sentiment_dict[word] = int(score)

File: maven-simplest/src/main/java/ETL/TextProcessor.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;

public class TextProcessor {
	private static HashMap<String, String> bannedWordsMap = new HashMap<String, String>();
	private static HashMap<String, Integer> sentimentScoresMap = new HashMap<String, Integer>();
	private static boolean bInitialized = false;

	private void Init() {
		String line = null;
		try {
			/* Banned words */
			FileReader fileReader = new FileReader("banned.txt");
			BufferedReader bufferedReader = new BufferedReader(fileReader);

			while ((line = bufferedReader.readLine()) != null) {
				String bannedWord = ROT13(line);
				String censoredWord = new String(new char[bannedWord.length() - 2]).replace('\0', '*');
				bannedWordsMap.put(bannedWord, censoredWord);
			}
			bufferedReader.close();
			/*
			 * for (String key: bannedWordsMap.keySet()) {
			 * System.out.println(key+": " + bannedWordsMap.get(key)); }
			 */
			
			/* Sentiment scores */
			fileReader = new FileReader("afinn.txt");
			bufferedReader = new BufferedReader(fileReader);

			while ((line = bufferedReader.readLine()) != null) {
				String[] tmp = line.split("\t", 2);
				String word = tmp[0];
				String score = tmp[1];
				sentimentScoresMap.put(word, Integer.parseInt(score));
			}
			bufferedReader.close();

			/*
			 * for (String key: sentimentScoresMap.keySet()) {
			 * System.out.println(key+": " + sentimentScoresMap.get(key)); }
			 */
		} catch (FileNotFoundException ex) {
			System.out.println("File 'banned.txt' Not Found!");
		} catch (IOException ex) {
			ex.printStackTrace();
		}
		bInitialized = true;
	}

	public TextProcessor() {
		if (!bInitialized) {
			Init();
		}
	}

	public String ROT13(String text) {
		String result = "";
		for (char ch : text.toCharArray()) {
			if (Character.isUpperCase(ch)) {
				ch = (char) ((ch - 'A' + 13) % 26 + 'A');
			} else if (Character.isLowerCase(ch)) {
				ch = (char) ((ch - 'a' + 13) % 26 + 'a');
			}
			result += ch;
		}
		return result;
	}

	public int SentimentScore(String text) {
		int score = 0;
		String[] words = text.split("[^\\w]+");
		for (int i = 0; i < words.length; i++) {
			String word = words[i].toLowerCase();
			if (sentimentScoresMap.containsKey(word)) {
				score += sentimentScoresMap.get(word);
			}
		}
		return score;
	}

	public String TextCensor(String text) {
		String cencoredText = text;
		String[] words = text.split("[^\\w]+");
		for (int i = 0; i < words.length; i++) {
			String word = words[i].toLowerCase();
			if (bannedWordsMap.containsKey(word)) {
				String cencoredWord = words[i].charAt(0) + bannedWordsMap.get(word)
						+ words[i].charAt(word.length() - 1);
				cencoredText = cencoredText.replaceAll("\\b" + words[i] + "\\b", cencoredWord);
			}
		}
		return cencoredText;
	}
}
